pkg/logger/server.go

```diff
@@ -0,0 +1,141 @@
+/*
+Copyright 2019 kubeflow.org.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package logger
+
+import (
+	"bytes"
+	"fmt"
+	"github.com/go-logr/logr"
+	guuid "github.com/google/uuid"
+	"github.com/kubeflow/kfserving/pkg/apis/serving/v1alpha2"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+)
+
+type LoggerHandler struct {
+	log       logr.Logger
+	svcHost   string
+	svcPort   string
+	logUrl    *url.URL
+	sourceUri *url.URL
+	logMode   v1alpha2.LoggerMode
+	modelId   string
+}
+
+func New(log logr.Logger, svcHost string, svcPort string, logUrl *url.URL, sourceUri *url.URL, logMode v1alpha2.LoggerMode, modelId string) http.Handler {
+	return &LoggerHandler{
+		log:       log,
+		svcHost:   svcHost,
+		svcPort:   svcPort,
+		logUrl:    logUrl,
+		sourceUri: sourceUri,
+		logMode:   logMode,
+		modelId:   modelId,
+	}
+}
+
+func (eh *LoggerHandler) callService(b []byte, r *http.Request) ([]byte, *string, error) {
+	url := &url.URL{
+		Scheme: "http",
+		Host:   fmt.Sprintf("%s:%s", eh.svcHost, eh.svcPort),
+		Path:   r.URL.Path,
+	}
+	eh.log.Info("Calling server", "url", url.String())
+	response, err := http.Post(url.String(), r.Header.Get("Content-Type"), bytes.NewReader(b))
+	if err != nil {
+		return nil, nil, fmt.Errorf("while calling post: %s", err)
+	}
+	rb, err := ioutil.ReadAll(response.Body)
+	if err != nil {
+		return nil, nil, fmt.Errorf("while reading response body: %s", err)
+	}
+	if err := response.Body.Close(); err != nil {
+		return nil, nil, fmt.Errorf("while closing response body: %s", err)
+	}
+	contentType := response.Header.Get("Content-Type")
+	return rb, &contentType, nil
+}
+
+func getOrCreateID(r *http.Request) string {
+	id := r.Header.Get("Ce-Id")
+	if id == "" {
+		id = guuid.New().String()
+	}
+	return id
+}
+
+// call svc and add send request/responses to logUrl
+func (eh *LoggerHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	// Read Payload
+	b, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		eh.log.Error(err, "Failed to read request payload")
+	}
+
+	// Get or Create an ID
+	id := getOrCreateID(r)
+
+	// log Request
+	if eh.logMode == v1alpha2.LogAll || eh.logMode == v1alpha2.LogRequest {
+		err = QueueLogRequest(LogRequest{
+			url:         eh.logUrl,
+			b:           &b,
+			contentType: "application/json", // Always JSON at present
+			reqType:     InferenceRequest,
+			id:          id,
+			sourceUri:   eh.sourceUri,
+			modelId:     eh.modelId,
+		})
+		if err != nil {
```

Comment:
Any reason to not do this inline?

if err :=  QueueLogRequest(LogRequest{
			url:         eh.logUrl,
			b:           &b,
			contentType: "application/json", // Always JSON at present
			reqType:     InferenceRequest,
			id:          id,
			sourceUri:   eh.sourceUri,
			modelId:     eh.modelId,
		}); err != nil {
...
}